Allow geometry selection independent of VecGeom being enabled

sethrj opened the pull request:
This does something similar with geometry that we have for the RNGs: selecting in-house geometry while still building the others. I'm thinking about adding a "simple ORANGE" for performance testing, and it'll be good to have it available without a branch.

This replaces a lot of `#if CELERITAS_USE_VECGEOM` with `#if CELERITAS_GEO == CELERITAS_GEO_VECGEOM`, which I think makes it a little clearer that it's about the main geometry selection rather than the availability of the external library.

I also renamed `load_gdml` to make it clearer that it's loading into the global geant4 state as opposed to VecGeom. The way it also tried to handle memory was also wrong: the GDML reader isn't actually returning a new'd pointer with C++ style memory management; the underlying object saves `this` to a magical vector that manages memory.

I also reworked the VecGeom test (since it can now be enabled without VecGeom being selected as the "core" geometry) so that it can run all the Geant4 geometry tests without having to break them apart into separate test instances or loading geant physics.

sethrj committed:
Genericize options and macros used for RNG, updating names for correct case

sethrj committed:
Define unavailable options as zero

sethrj committed:
Add compile-time selection of ORANGE even when VecGeom is being built

sethrj committed:
Use string_view for test

sethrj committed:
Fix memory management in GeantSetup and add helpers for cleaning geometry store

sethrj committed:
Allow Geant4->VecGeom tests to run without RunManager

sethrj committed:
Fix warning about air having zero density

sethrj committed:
Improve logging

sethrj committed:
Clean up some issues with "vecgeom enabled" vs "vecgeom default"

sethrj committed:
Rename `load_gdml` and file

sethrj committed:
Clean implicit copy constructor warnings

sethrj committed:
Fix xorwow spelling and add geo to build output

sethrj committed:
Fix no-vecgeom build

sethrj committed:
Disable rasterizer test when vecgeom isn't the core geometry

mrguilima commented on cmake/CeleritasUtils.cmake:
```diff
@@ -245,6 +255,20 @@ endfunction()
 #-----------------------------------------------------------------------------#
 
+macro(celeritas_setup_option var option) #[condition]
+  if(${ARGC} EQUAL 2)
+    # always-on-option
+    list(APPEND ${var}_OPTIONS ${option})
+  elseif(${ARGV2})
+    # variable evaluates to true
+    list(APPEND ${var}_OPTIONS ${option})
```
Are those two branches supposed to have identical bodies?
Or is this a copy-and-paste oversight?

sethrj replied:
They are supposed to be identical, and I *think* it's necessary because this is a cmake "macro" rather than a "function"...

pcanal commented on cmake/CeleritasUtils.cmake:
```diff
@@ -64,6 +64,39 @@ CMake configuration utility functions for Celeritas.
+
+    celeritas_define_options(<var> <doc>)
+
+  This will set <var> to the first item of the list ``${<var>_OPTIONS}``. It
+  will validate that the selection is one of the list and default to the first
+  item in the list.
```
```suggestion
  This will set <var> to the first item of the list ``${<var>_OPTIONS}``.
```
Aren't the next two sentences 'redundant' (Since it is the first element in the list it is guaranteed to be in the list)

pcanal replied:
Reading the actual code.  The additional sentences are not redundant, just confusing.
```suggestion
  If <var> is not yet set, this will set it to the first item of the list ``${<var>_OPTIONS}``. 
  Otherwise It will validate that the pre-existing selection is one of the list.
```

pcanal commented on cmake/CeleritasUtils.cmake:
```diff
@@ -64,6 +64,39 @@ CMake configuration utility functions for Celeritas.
+
+  Generate preprocessor macros for the given option list.::
+
+    celeritas_generate_option_macros(<var>)
```
```suggestion
    celeritas_generate_option_definitions(<var>)
```
CMake also has the concept of [macros](https://cmake.org/cmake/help/latest/command/macro.html) and thus this was my first thought (until I read the documentation with indeed talks about `C preprocessor` in passing.  

The propose change is keeping in tone with CMake itself, for example [add_definition](https://cmake.org/cmake/help/latest/command/add_definitions.html)

pcanal commented on cmake/CeleritasUtils.cmake:
```diff
@@ -64,6 +64,39 @@ CMake configuration utility functions for Celeritas.
+  This requires the list of ``<var>_OPTIONS`` to be set and ``<var>`` to be set,
+  and it creates a string in the parent scope called ``<var>_MACROS``
+
+  The resulting macro list starts a counter counter from 1 because undefined
+  macros have the implicit value of 0 in the C preprocessor. Thus any
+  unavailable options (e.g. CELERITAS_USE_CURAND when HIP is in use) will
+  implicitly be zero.
```
```suggestion
  The resulting macro list starts its counter at 1 since undefined C macros have 
  the implicit value of 0 in the C preprocessor. Thus any unavailable options (e.g. 
  CELERITAS_USE_CURAND when HIP is in use) will implicitly be zero.
```

pcanal replied:
I also suggest to add:
```
    The resulting variable will be used when generating the config header file (usually
    `celeritas_config.h`) which needs to contain `@<var>_MACROS@` 
```
(and independently `_MACROS` could be renamed `_DEFINITION` )

sethrj replied:
I think `_CONFIG` might even be clearer than macros or definition; so I've renamed it accordingly.

pcanal replied:
I think so too. Thanks.

pcanal commented on cmake/CeleritasUtils.cmake:
```diff
@@ -221,3 +254,66 @@ function(celeritas_configure_file input output)
+
+  list(FIND ${var}_OPTIONS "${${var}}" _index)
+  if(_index EQUAL -1)
+    message(SEND_ERROR "Invalid value ${var}=${${var}}: must be "
+      "${${var}_OPTIONS}; overriding for next configure")
```
Consider making this fatal rather than 'fixing' it.  If for some reason the message is 'lost' and/or missed by the user (eg. terminal windows dies after cmake finished but before the user reads it or lost in a flood or errors), they will not notice that the build is working but with an option that is (arbitrarily) different from their selection.

sethrj replied:
Yeah, I think this happened to Guilherme... I like the idea that successive runs of cmake will resolve conflicting configure options, but I see what you mean. I'll think about it.

sethrj replied:
OK @pcanal , now this prints an error message:
```
CMake Error at cmake/CeleritasUtils.cmake:268 (message):
  Invalid selection not in list: "ORANGE" (CELERITAS_GEO=VecGeom): run `cmake
  -DCELERITAS_GEO=ORANGE` for a possible fix
```

We could make it smarter by querying the cache variable properties to maybe deduce if the variable was set by the user (don't override) versus an internal default (do override) but that can be deferred.

pcanal replied:
I'll like the proposed message (which seems to not be uploaded yet).  [Not clear from the example but I assume that the first quote would contain the user value (the current example contains a valid value)]

sethrj replied:
It should be there: https://github.com/celeritas-project/celeritas/pull/726/commits/a80a5f7e35ecb39b5835bd593aa058827e6e06f7: the quoted is the list (which in this case has only one entry, so not sure what to do with that). The actual value is in parens.

pcanal replied:
ok. I see.  Maybe consider this variation:
```
The selection "CELERITAS_GEO=VecGeom" is invalid; 
   the value must be in the following list: "ORANGE";
   run `cmake -DCELERITAS_GEO=ORANGE` for a possible fix
```

sethrj replied:
Unfortunately I can't do that while reusing the macro across all the "invalid option" cases... how about:
```
CMake Error at cmake/CeleritasUtils.cmake:268 (message):
  Invalid option for CELERITAS_CORE_GEO: must be in the list [ORANGE]
  (CELERITAS_CORE_GEO=VecGeom): run `cmake -DCELERITAS_CORE_GEO=ORANGE` for a
  possible fix
```

pcanal replied:
> I can't do that while reusing the macro across all the "invalid option" cases

~I don't quite understand. I thought my proposal was only a reordering of the words and variables.  What am I missing?~  I see there is a function now ... let me see.

sethrj replied:
The macro `celeritas_error_incompatible_option`, which abstracts the decision to change parameter vs warning , displays the variable name and its value after a prespecified message:
```cmake
function(celeritas_error_incompatible_option msg var new_value)
  message(SEND_ERROR "${msg} (${var}=${${var}}):"
    "  run `cmake -D${var}=${new_value}` for a possible fix"
  )
endfunction()
```
remember that this macro is also used in a couple other places:
```cmake
  if(NOT Doxygen_FOUND)
    celeritas_error_incompatible_option(
      "Doxygen is required for building documentation but was not found"
      "CELERITAS_BUILD_DOCS" OFF
    )
  endif()
```
and
```cmake
    celeritas_error_incompatible_option(
      "CUDA mismatch between the VecGeom installation "
      "at ${VECGEOM_INSTALL_DIR} and Celeritas"
      CELERITAS_USE_CUDA
      "${VecGeom_CUDA_FOUND}"
    )
```

so I can't simply reorder the message like you asked.

pcanal replied:
The wording @sethrj propose at https://github.com/celeritas-project/celeritas/pull/726#issuecomment-1523470906 in conjunction with the update https://github.com/celeritas-project/celeritas/pull/726/files#r1177952200 would means a text like:
```
CMake Error at cmake/CeleritasUtils.cmake:268 (message):
  Invalid option for CELERITAS_CORE_GEO: must be in the list [ORANGE]
  (This is caused by CELERITAS_CORE_GEO being set to VecGeom): 
  run `cmake -DCELERITAS_CORE_GEO=ORANGE` for a possible fix
```
looks clear enough to me.

pcanal replied:
```
< Invalid option for CELERITAS_CORE_GEO: must be in the list [ORANGE]
> Invalid option for CELERITAS_CORE_GEO, it must be in the list [ORANGE]
```

sethrj replied:
Your message has a comma splice 😉 and I prefer error messages not to be so prosaic. I think this should satisfy us both:
```
CMake Error at cmake/CeleritasUtils.cmake:268 (message):
  Invalid setting CELERITAS_CORE_GEO=VecGeom: valid options are {ORANGE}

      Possible fix: cmake -DCELERITAS_CORE_GEO=ORANGE /Users/seth/.local/src/celeritas-temp/build
Call Stack (most recent call first):
  cmake/CeleritasUtils.cmake:302 (celeritas_error_incompatible_option)
  CMakeLists.txt:120 (celeritas_define_options)
```

pcanal replied:
> a comma splice 

A period would do just fine then :)

pcanal replied:
>  I think this should satisfy us both:

Yes, I agree :)

pcanal commented on cmake/CeleritasUtils.cmake:
```diff
@@ -221,3 +254,66 @@ function(celeritas_configure_file input output)
+  set(_options)
+  foreach(_val IN LISTS ${var}_DISABLED_OPTIONS)
+    string(TOUPPER "${_val}" _val)
+    list(APPEND _options "#define ${var}_${_val} 0")
```
A more verbose approach (resulting in a more typical use where both `#if not defined(..)` and `#if not ..` work
```suggestion
    list(APPEND _options "#if defined(${var}_${_val})")
    list(APPEND _options "#undef ${var}_${_val} 0")
    list(APPEND _options "#endif")
```
with this
```
#if SELECTION == UNDEFINED_VAR
...
#endif
```
still works as expected (correctly skip over the section and completes without errors)

sethrj replied:
Yeah but then you can't use it as a C++ constant... so you won't be able to write `if (CELERITAS_GEO == CELERITAS_GEO_VECGEOM)`.

pcanal replied:
In what context would that make sense (i.e. where the code after the if does not need any vecgeom symbols)?

sethrj replied:
The most common one is for test results that change depending on the selection of the "core" geometry: see https://github.com/celeritas-project/celeritas/pull/726/files#diff-a40e865ce029c1d7d0e6db09a86cb5feb80985417e1b89798194377afc4d0b1e

pcanal replied:
I see.  It does make sense.

pcanal commented on app/CMakeLists.txt:
```diff
@@ -173,19 +173,22 @@ endif()
     nlohmann_json::nlohmann_json
-    VecGeom::vecgeom
   )
+  if(CELERITAS_USE_CUDA AND CELERITAS_DEBUG AND CELERITAS_GEO STREQUAL "VecGeom")
```
```suggestion
  if(CELERITAS_GEO STREQUAL "VecGeom")
```
as the statement is already is a `CELERITAS_USE_CUDA is true` section and I don't see why you need the `VecGeom` libraries only in debug mode.

pcanal commented on src/celeritas/CMakeLists.txt:
```diff
@@ -130,8 +130,6 @@ if(CELERITAS_USE_Geant4)
       ext/detail/GeantGeoConverter.cc
     )
     list(APPEND _cg4_libs VecGeom::vecgeom)
-  else()
```
On line 128,(might not be visible on github in the default view) there is:
```
if(CELERITAS_USE_VecGeom)
```
should we consistently use:
```
if(CELERITAS_GEO STREQUAL "VecGeom")
```
instead or is there a different semantic here?

sethrj replied:
Yeah, it's a different semantic: this changeset will allow us to build VecGeom but make it not the "core" geometry. In practice I don't think we'll want to do this, but it's sort of the first step toward potentially having core components templated on the geometry type.

pcanal replied:
In that case, I think it might be clearer to rename the old one `CELERITAS_HAS_VecGeom`

sethrj replied:
USE, HAS, HAVE ... we're still "using" it (compiling it) but not necessarily using it for everything. You could say the same thing about CUDA or MPI.

pcanal replied:
I am trying to make a distinction between:
```
CELERITAS_GEO STREQUAL "VecGeom"   -> Used as the main geometry backend
and
CELERITAS_HAS_VecGeom -> Is available/compiled-in (and might be used auxiliary code).
```
(In my humble opinion) "use" implied that it is the utilized from the main part of the project (i.e. here, as a geometry backend for the main Celeritas library".  (And also if the code becomes completely templated on the geometry backend, then which is ever is being "used" is now per se delegated to the user code/project/library)

pcanal replied:
>  You could say the same thing about CUDA or MPI.

The difference is that (at least for now :) ) we don't have 2 closely related `CMake` variable that can easily be confused one for the other (esp. since the `use` one used to have/include the semantic of the new `_GEO` variable).

sethrj replied:
I don't want to rename all the `CELERITAS_USE_([A-Z]+)` to `CELERITAS_HAVE_\1`. Do you think it's worth renaming `CELERITAS_(RNG|GEO)` to `CELERITAS_CORE_\1`? We do compile CUDA/HIP RNGs if available even though we might be using the in-house XORWOW.

pcanal replied:
> I don't want to rename all

I agree as all the other are unambiguous.  I was recommending to **only** change the  `CELERITAS_USE_*` to `CELERITAS_HAS_\1`.

But I guess, it would also make sense for the `RNG` ones.

sethrj replied:
It doesn't make any sense to have `CELERITAS_HAS_VecGeom` and `CELERITAS_USE_CUDA` doing analogous thing: showing that an optional library is being compiled in.

mrguilima replied:
Is the CELERITAS_HAS_VecGeom a CMake-level configuration or a C++ macro?
If VecGeom was built into the executable, can the geometry system be chosen at run-time as well, or at build-time only?

sethrj replied:
Both. And for now it's build-time, but if we add templates to everything that uses the CoreTrackData, CoreTrackView, or GeoTrackView, it could be a runtime. Our MC code Shift does this but I deliberately said we should avoid that for now for the sake of simplicity.

pcanal replied:
> It doesn't make any sense to have CELERITAS_HAS_VecGeom and CELERITAS_USE_CUDA doing analogous thing: 

That's a good point ... so indeed it would be all or nothing.

> Do you think it's worth renaming `CELERITAS_(RNG|GEO)` to `CELERITAS_CORE_\1`?

maybe ... I guess (eventually) it would be also semantically close to `CELERITAS_DEFAULT_\1`  ....

sethrj replied:
`CELERITAS_DEFAULT_` would I think become `CELERITAS_CORE_DEFAULT_`...

pcanal replied:
Sounds good to me :)

sethrj committed:
Incorporate @pcanal feedback

sethrj committed:
By default don't change the cmake variables

Provide a helper function that we can use to centralize the
decision/implementation between outputting a message or changing the
function.

sethrj committed:
Merge remote-tracking branch 'upstream/develop' into geo-choices

sethrj committed:
Use CELERITAS_CORE_(RNG|GEO) for clarity

sethrj committed:
Improve error messages

sethrj commented:
@pcanal Any further comments? I think I've addressed all your concerns. If you want a clearer error message perhaps the following is good enough:
```diff
diff --git a/cmake/CeleritasUtils.cmake b/cmake/CeleritasUtils.cmake
index b70f8abc..77d3d304 100644
--- a/cmake/CeleritasUtils.cmake
+++ b/cmake/CeleritasUtils.cmake
@@ -298,8 +298,9 @@ function(celeritas_define_options var doc)

   list(FIND ${var}_OPTIONS "${${var}}" _index)
   if(_index EQUAL -1)
+    string(JOIN "," _optlist ${${var}_OPTIONS})
     celeritas_error_incompatible_option(
-      "Invalid selection not in list: \"${${var}_OPTIONS}\""
+      "Invalid option for ${var}: must be in the list [${_optlist}] "
       "${var}" "${_default}"
     )
   endif()
```

pcanal commented on cmake/CeleritasUtils.cmake:
```diff
@@ -221,3 +263,75 @@ function(celeritas_configure_file input output)
+
+function(celeritas_error_incompatible_option msg var new_value)
+  message(SEND_ERROR "${msg} (${var}=${${var}}):"
+    "  run `cmake -D${var}=${new_value}` for a possible fix"
```
```suggestion
  message(SEND_ERROR "${msg} (This is caused by ${var} being set to ${${var}}):"
    "  run `cmake -D${var}=${new_value}` for a possible fix"
```

pcanal reviewed (approved):
Thanks
